elmslie-core/src/main/java/vivid/money/elmslie/core/store/ElmStore.kt

```diff
@@ -1,136 +1,137 @@
+    private val effectBuffer = ConcurrentLinkedQueue<Effect>()
+    private val effectBufferingListener = effectBuffer::add
+    private val effectListeners = ConcurrentHashSet<(Effect) -> Any?>()
+    private val eventListeners = ConcurrentHashSet<(Event) -> Any?>()
+    private val stateListeners = ConcurrentHashSet<(State) -> Any?>()
+    private val stateInternal = AtomicReference(initialState)
+    override val currentState: State get() = stateInternal.get()
 
-    override fun start(): Store<Event, Effect, State> {
-        if (!isStartedInternal.compareAndSet(false, true)) {
-            logger.fatal(
-                "Store start error",
-                Exception("Store is already started. Usually, it happened inside StoreHolder.")
-            )
-        }
-        effectsBuffer.init().bind()
+    // We can't use subject to store state to keep it synchronized with children
+    private val stateLock = Any()
 
-        eventsInternal
-            .observeOn(computation())
-            .flatMap { event ->
-                logger.debug("New event: $event")
-                val (effects, commands) = synchronized(stateReferenceLock) {
-                    val state = statesInternal.value!!
-                    val result = reducer.reduce(event, state)
-                    statesInternal.onNext(result.state)
-                    result.effects to result.commands
-                }
-                effects.forEach { logger.debug("New effect: $it") }
-                effects.forEach(effectsInternal::onNext)
-                Observable.fromIterable(commands)
-            }
-            .subscribe(commandsInternal::onNext) {
-                logger.fatal("You must handle all errors inside reducer", it)
-            }
-            .bind()
-
-        commandsInternal
-            .flatMap { command ->
-                logger.debug("Executing command: $command")
-                actor.execute(command)
-                    .subscribeOn(io())
-                    .doOnError { logger.nonfatal(error = it) }
-                    .onErrorResumeNext { Observable.empty() }
-            }
-            .subscribe(eventsInternal::onNext) {
-                logger.fatal("Unexpected actor error", it)
-            }
-            .bind()
+    override fun accept(event: Event) = dispatchEvent(event)
 
-        return this
+    override fun start() = this.also {
+        requireNotStarted()
+        stopBuffering()
     }
 
     override fun stop() {
         isStartedInternal.set(false)
         disposables.clear()
+        startBuffering()
+    }
+
+    override fun states(onStateChange: (State) -> Unit): Disposable {
+        val callback = onStateChange.distinctUntilChanged()
+        stateListeners += callback
+        dispatchState(currentState)
+        return Disposable { stateListeners -= callback }
+    }
+
+    override fun effects(onEffectEmission: (Effect) -> Unit): Disposable {
+        dispatchBuffer(onEffectEmission)
+        startBuffering()
```

Comment:
Yes, this is the intended behavior, which is reflected in the method documentation. There are very little usecases for this method and as far as i understand this behavior supports all of them.

Reply:
1. send some effects and wait while it will be shown
2. put the application in the background and then get it
3. all events will be shown again.
I tried this on the current branch. Is this intended behavior?

Reply:
@DemD10, I've tried to reproduce this on sample-android-loader and the duplicate effects don't appear.  Which sample can I use to receive duplicate effects?

Reply:
@DemD10 I was able to repoduce and fix this in [this](https://github.com/vivid-money/elmslie/pull/98/commits/558b7d3b27f7e22a58e2a8b8bd498e451510bb15) commit